### DISCUSSION TITLE
Understanding EM/EB and codes

### DISCUSSION THREAD

**Discussion by nickaws** [ki3v/xtramtest#7] [General]
Hi
what does EM and EB mean?

I am getting the hex number under EM in the addresses from 256k-320k on an IBM 5150 with an expansion card. I don't get anything under EB. The video from Adrian doesn't seem to have that case covered

Thank you,

**Comment by ki3v** [comment]
EM is errors from the March test, and EB is errors from the bit pattern (actually Ganssle bus exercise) test. 

The hex code should tell you which chip in that bank is giving you trouble. Because there is an error under March (EM) but not Ganssle (EB), it suggests a page error inside the chip. Of course it doesn’t really matter too much what the kind of error is… a bad chip is a bad chip.

Also note that if an address pin of a memory chip is broken or not seated in its socket, or there’s a broken trace or cold solder joint on an address pin (in other words, not making a connection) it can give the same result, even if the DRAM chip is good. Just something to keep in mind. 

Please let me know if you are able to successfully repair the board, or if you continue to have trouble.

**Comment by nickaws** [comment]
Hi! thanks for your answer, this makes a lot of sense. Unfortunately, the memory bank that is problematic is soldered, so at then moment there's not a reasonable way for me to fix, tried to remove one chip... and lifted one pad, so I will need to wait until I get a desoldering gun.

I saw an interesting behaviour, however. I removed all the chips except the bank that is problematic

Sometimes it marks 90 under the the column, sometimes marks 80. Sometimes marks 90 and the last row is 80.

I understand if all are 80 or all are 90, there's a chip that is just bad sometimes, however what does it mean when the last one (4c-4f) is 80? I don't have a photo, haven't been able to replicate, but it's real - not that it particularly matters, just curiours

![image](https://github.com/user-attachments/assets/8649bd2c-56b1-4820-a61d-1271bbb20e0b)

**Comment by nickaws** [comment]
Not an actual issue, just some extra observations.

By the way, I used the piggyback technique, and when I placed two new chips into the correct positions (10010000), I encountered another faulty chip (00001000). Once I added a third chip, the test no longer found any errors after multiple passes.

I also have the parity error chip with issues, but the BIOS identified that one as well.

**Comment by ki3v** [comment]
Regarding one line sometimes reading 80 instead of 90, I'm guessing that your intuition would be that one of the chips just sometimes doesn't show up as errored when testing in that region. That's my intuition too. As much as I like March tests, they reveal _where_ errors are, but not _which_ kind of error it is. 

Also interesting about the piggyback method causing an error on a different chip/bit. Assuming these are 1-bit wide DRAMs (8/9 per bank) it could be that one of the piggyback chips is loading the address bus down (changing voltage or timing slightly) enough that one other chip, which has marginal address logic and starts to misbehave. That's a guess though.

**Comment by nickaws** [comment]
Here's another strange occurrence from when I rebooted the machine:

The F6 changed to FF later, and I have no chips installed in the 320k-640k range.

As you suggested, this might also indicate a problem with the card misbehaving. I will update you once I get the desoldering gun, replace the chips, and see how it behaves.

I appreciate all your input on this.



![WhatsApp Image 2024-09-13 at 11 24 59 AM](https://github.com/user-attachments/assets/8609f9b5-05dc-4e40-9ad0-dccd1ff80c81)

**Comment by ki3v** [comment]
That F6 is very odd indeed.  I wonder if there could be a problem with some chip select logic on the board?  F6 suggests that the two least significant bits are responding correctly (at least under the non-March test), which implies that some chip is acting on reads and writes when those addresses are accessed.

Incidentally, that Mitsubishi monitor reminds me very much of one of the early monitors I had as a teenager on my first PC.

**Comment by mskipatrol** [comment]
I ran into a similar scenario with an XT Clone board from a Sanyo MBC-885.
I have no documentation on this board so I wasn't clear on the memory chip layout.
Also all memory chips are soldered on this board which makes replacing them a pain.
I was getting error 80 in all positions on the second bank, EM and EB. I had tried
replacing the Bit 0 and Bit 7 chips (green) in what I thought to be bank 2 with no success.
Tried the piggyback method on several chips and noticed that one chip in particular
in what I thought was bank 4 when piggybacked removed several of the error 80's.
Added two additional piggyback chips in that same bank, which cleared up all errors.
The first chip I had tried seemed to make the most difference, so I replaced that chip
(circled in red) and that cleared all the errors. It seems that adding those extra two 
piggyback chips loaded the memory bus allowing things to work, but in the end it was 
just the one chip causing the errors.  

<img width="4000" height="2250" alt="Error 80" src="https://github.com/user-attachments/assets/fc1c3d86-529c-491b-899f-7b787135e61a" />

<img width="4000" height="2250" alt="Replaced chips" src="https://github.com/user-attachments/assets/96c98294-1ff3-4139-ab5a-2b732ff2f36b" />
